File: approximations/finder/src/main/java/com/wittsfamily/approximations/finder/cleaning/CleaningAdd.java
package com.wittsfamily.approximations.finder.cleaning;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

import com.wittsfamily.approximations.finder.ExpressionRenderer;

public class CleaningAdd implements CleaningExpression {
    private List<CleaningExpression> expressions = new ArrayList<>();
    private CleaningExpression parent;

    public void addExpression(CleaningExpression child) {
        expressions.add(child);
        child.setParent(this);
    }

    @Override
    public List<CleaningExpression> getExpressions() {
        return expressions;
    }

    @Override
    public CleaningExpressionType getType() {
        return CleaningExpressionType.ADD;
    }

    @Override
    public CleaningExpression getParent() {
        return parent;
    }

    @Override
    public void setParent(CleaningExpression parent) {
        this.parent = parent;
    }

    @Override
    public boolean cleanOne(boolean convertPowers) {
        boolean hasDoneAnything = false;
        for (Iterator<CleaningExpression> iterator = expressions.iterator(); iterator.hasNext();) {
            CleaningExpression child = iterator.next();
            if (child.getType() == CleaningExpressionType.INTEGER && ((CleaningInteger) child).asInt() == 0) {
                iterator.remove();
                hasDoneAnything = true;
            }
        }
        if (expressions.size() == 1) {
            parent.replaceChild(this, expressions.get(0));
            return true;
        } else if (expressions.size() == 0) {
            parent.replaceChild(this, new CleaningInteger(0));
            return true;
        }
        for (int i = 0; i < expressions.size(); i++) {
            CleaningExpression exp = expressions.get(i);
            hasDoneAnything = exp.cleanOne(convertPowers) || hasDoneAnything;
        }
        return hasDoneAnything;
    }

    @Override
    public boolean cleanTwo() {
        boolean hasDoneAnything = false;
        for (int i = 0; i < expressions.size(); i++) {
            CleaningExpression exp = expressions.get(i);
            hasDoneAnything = exp.cleanTwo() || hasDoneAnything;
        }
        return hasDoneAnything;
    }

    @Override
    public boolean cleanThree() {
        boolean hasDoneAnything = false;
        for (int i = 0; i < expressions.size(); i++) {
            CleaningExpression exp = expressions.get(i);
            hasDoneAnything = exp.cleanThree() || hasDoneAnything;
        }
        List<CleaningLog> logs = new ArrayList<>();
        List<CleaningNegate> negLogs = new ArrayList<>();
        for (CleaningExpression child : expressions) {
            if (child.getType() == CleaningExpressionType.LOG) {
                logs.add((CleaningLog) child);
            } else if (child.getType() == CleaningExpressionType.NEGATE && child.getExpressions().get(0).getType() == CleaningExpressionType.LOG) {
                negLogs.add((CleaningNegate) child);
            }
        }
        List<CleaningExpression> toRemove = new ArrayList<>();
        CleaningExpression newChild = null;
        for (CleaningLog log : logs) {
            List<CleaningLog> matchingLogs = new ArrayList<>();
            List<CleaningNegate> matchingNegLogs = new ArrayList<>();
            CleaningExpression base = log.getBase();
            for (CleaningLog otherLog : logs) {
                if (otherLog.getBase().equals(base) && otherLog != log) {
                    matchingLogs.add(otherLog);
                }
            }
            for (CleaningNegate negLog : negLogs) {
                if (((CleaningLog) negLog.getExpressions().get(0)).getBase().equals(base)) {
                    matchingNegLogs.add(negLog);
                }
            }
            if (matchingLogs.size() + matchingNegLogs.size() > 1) {
                CleaningMultiply mult = new CleaningMultiply();
                toRemove.addAll(matchingLogs);
                toRemove.addAll(matchingNegLogs);
                for (CleaningLog l : matchingLogs) {
                    mult.addExpression(l.getChild());
                }
                for (CleaningNegate l : matchingNegLogs) {
                    mult.addExpression(new CleaningInvert(((CleaningLog) l.getExpressions().get(0)).getChild()));
                }
                newChild = new CleaningLog(base, mult);
                break;
            }
        }
        if (newChild == null) {
            for (CleaningNegate negate : negLogs) {
                List<CleaningNegate> matchingNegLogs = new ArrayList<>();
                CleaningExpression base = ((CleaningLog) negate.getExpressions().get(0)).getBase();
                for (CleaningNegate negLog : negLogs) {
                    if (((CleaningLog) negLog.getExpressions().get(0)).getBase().equals(base)) {
                        matchingNegLogs.add(negLog);
                    }
                }
                if (matchingNegLogs.size() > 1) {
                    CleaningMultiply mult = new CleaningMultiply();
                    toRemove.addAll(matchingNegLogs);
                    for (CleaningNegate l : matchingNegLogs) {
                        mult.addExpression(((CleaningLog) l.getExpressions().get(0)).getChild());
                    }
                    newChild = new CleaningNegate(new CleaningLog(base, mult));
                    break;
                }
            }

        }
        if (newChild != null) {
            hasDoneAnything = true;
            newChild.setParent(this);
            expressions.removeAll(toRemove);
            expressions.add(newChild);
        }
        return hasDoneAnything;
    }

    @Override
    public boolean cleanFour() {
        boolean didSomething = false;
        int totalIntVal = 0;
        List<CleaningReal> reals = new ArrayList<>();
        List<CleaningReal> negReals = new ArrayList<>();
        for (Iterator<CleaningExpression> iterator = expressions.iterator(); iterator.hasNext();) {
            CleaningExpression child = iterator.next();
            if (child.getType() == CleaningExpressionType.INTEGER) {
                totalIntVal += ((CleaningInteger) child).asInt();
                iterator.remove();
            } else if (child.getType() == CleaningExpressionType.REAL) {
                reals.add((CleaningReal) child);
                iterator.remove();
            } else if (child.getType() == CleaningExpressionType.MULTIPLY && child.getExpressions().size() == 2) {
                if (child.getExpressions().get(0).getType() == CleaningExpressionType.INTEGER && child.getExpressions().get(1).getType() == CleaningExpressionType.REAL) {
                    int num = ((CleaningInteger) child.getExpressions().get(0)).asInt();
                    for (int i = 0; i < num; i++) {
                        reals.add((CleaningReal) child.getExpressions().get(1));
                    }
                    iterator.remove();
                } else if (child.getExpressions().get(1).getType() == CleaningExpressionType.INTEGER && child.getExpressions().get(0).getType() == CleaningExpressionType.REAL) {
                    int num = ((CleaningInteger) child.getExpressions().get(1)).asInt();
                    for (int i = 0; i < num; i++) {
                        reals.add((CleaningReal) child.getExpressions().get(0));
                    }
                    iterator.remove();
                }
            } else if (child.getType() == CleaningExpressionType.NEGATE) {
                CleaningExpression cc = child.getExpressions().get(0);
                if (cc.getType() == CleaningExpressionType.INTEGER) {
                    totalIntVal -= ((CleaningInteger) cc).asInt();
                    iterator.remove();
                } else if (cc.getType() == CleaningExpressionType.REAL) {
                    negReals.add((CleaningReal) cc);
                    iterator.remove();
                } else if (cc.getType() == CleaningExpressionType.MULTIPLY && cc.getExpressions().size() == 2) {
                    if (cc.getExpressions().get(0).getType() == CleaningExpressionType.INTEGER && cc.getExpressions().get(1).getType() == CleaningExpressionType.REAL) {
                        int num = ((CleaningInteger) cc.getExpressions().get(0)).asInt();
                        for (int i = 0; i < num; i++) {
                            negReals.add((CleaningReal) cc.getExpressions().get(1));
                        }
                        iterator.remove();
                    } else if (cc.getExpressions().get(1).getType() == CleaningExpressionType.INTEGER && cc.getExpressions().get(0).getType() == CleaningExpressionType.REAL) {
                        int num = ((CleaningInteger) cc.getExpressions().get(1)).asInt();
                        for (int i = 0; i < num; i++) {
                            negReals.add((CleaningReal) cc.getExpressions().get(0));
                        }
                        iterator.remove();
                    }
                }
            }
        }
        if (totalIntVal != 0) {
            if (totalIntVal > 0) {
                expressions.add(new CleaningInteger(totalIntVal));
            } else {
                expressions.add(new CleaningNegate(new CleaningInteger(-totalIntVal)));
            }
        }
        for (Iterator<CleaningReal> iterator = reals.iterator(); iterator.hasNext();) {
            CleaningReal r = iterator.next();
            if (negReals.contains(r)) {
                negReals.remove(r);
                iterator.remove();
                didSomething = true;
            }
        }
        while (reals.size() > 0) {
            CleaningReal r = reals.get(0);
            reals.remove(0);
            int num = 1;
            for (int i = 0; i < reals.size(); i++) {
                if (reals.get(i).equals(r)) {
                    reals.remove(i--);
                    num++;
                    didSomething = true;
                }
            }
            for (int i = 0; i < negReals.size(); i++) {
                if (negReals.get(i).equals(r)) {
                    negReals.remove(i--);
                    num--;
                    didSomething = true;
                }
            }
            if (Math.abs(num) == 1) {
                if (num > 0) {
                    expressions.add(r);
                } else {
                    CleaningNegate n = new CleaningNegate(r);
                    n.setParent(this);
                    expressions.add(n);
                }
            } else if (num != 0) {
                CleaningMultiply mult = new CleaningMultiply();
                mult.addExpression(r);
                if (num > 0) {
                    mult.addExpression(new CleaningInteger(num));
                    mult.setParent(this);
                    expressions.add(mult);
                } else {
                    mult.addExpression(new CleaningInteger(-num));
                    CleaningNegate n = new CleaningNegate(mult);
                    n.setParent(this);
                    expressions.add(n);
                }
            }
        }
        while (negReals.size() > 0) {
            CleaningReal r = negReals.get(0);
            negReals.remove(0);
            int num = -1;
            for (int i = 0; i < negReals.size(); i++) {
                if (negReals.get(i).equals(r)) {
                    negReals.remove(i--);
                    num--;
                    didSomething = true;
                }
            }
            if (Math.abs(num) == 1) {
                CleaningNegate n = new CleaningNegate(r);
                n.setParent(this);
                expressions.add(n);
            } else {
                CleaningMultiply mult = new CleaningMultiply();
                mult.addExpression(r);
                mult.addExpression(new CleaningInteger(-num));
                CleaningNegate n = new CleaningNegate(mult);
                n.setParent(this);
                expressions.add(n);
            }
        }
        for (int i = 0; i < expressions.size(); i++) {
            didSomething = expressions.get(i).cleanFour() || didSomething;
        }
        return didSomething;
    }

    @Override
    public String render(ExpressionRenderer r) {
        StringBuilder pos = new StringBuilder();
        StringBuilder neg = new StringBuilder();
        boolean hasPrev = false;
        for (CleaningExpression exp : expressions) {
            if (exp.getType() == CleaningExpressionType.NEGATE) {
                neg.append(exp.render(r));
            } else {
                if (hasPrev) {
                    pos.append(r.plus());
                }
                pos.append(exp.render(r));
                hasPrev = true;
            }
        }
        return pos.toString() + neg.toString();
    }

    @Override
    public void replaceChild(CleaningExpression oldExp, CleaningExpression newExp) {
        expressions.removeIf(c -> c == oldExp);
        expressions.add(newExp);
        newExp.setParent(this);
    }

    @Override
    public boolean equals(Object obj) {
        if (obj.getClass() != CleaningAdd.class) {
            return false;
        }
        CleaningAdd other = (CleaningAdd) obj;
        if (other.expressions.size() != expressions.size()) {
            return false;
        }
        if (other.expressions.containsAll(expressions)) {
            List<CleaningExpression> exps = new ArrayList<>(other.expressions);
            for (CleaningExpression child : expressions) {
                if (!exps.remove(child)) {
                    return false;
                }
            }
            return true;
        }
        return false;
    }

    @Override
    public boolean canNegate() {
        for (CleaningExpression child : expressions) {
            if (child.getType() == CleaningExpressionType.NEGATE) {
                return true;
            }
        }
        return false;
    }

    @Override
    public void negate() {
        List<CleaningExpression> newExpressions = new ArrayList<>();
        for (CleaningExpression child : expressions) {
            if (child.getType() == CleaningExpressionType.NEGATE) {
                newExpressions.add(child.getExpressions().get(0));
            } else {
                newExpressions.add(new CleaningNegate(child));
            }
        }
        for (CleaningExpression newExp : newExpressions) {
            newExp.setParent(this);
        }
        expressions = newExpressions;

    }

    @Override
    public double asDouble() {
        double d = 0;
        for (CleaningExpression exp : expressions) {
            d += exp.asDouble();
        }
        return d;
    }

    @Override
    public CleaningAdd clone() {
        CleaningAdd clone = new CleaningAdd();
        for (CleaningExpression child : expressions) {
            clone.addExpression(child.clone());
        }
        return clone;
    }

}

File: approximations/server-side-rest/src/main/resources/public/cleaningAdd.js
/**
 * 
 */

class CleaningAdd {
	constructor(){
		this.expressions = [];
		this.parent = null;
	}
	addExpression(child) {
		this.expressions.push(child);
		child.setParent(this);
	}

	getExpressions() {
		return this.expressions;
	}

	getType() {
		return CleaningExpressionType.ADD;
	}

	getParent() {
		return this.parent;
	}
	
	setParent(parent) {
		this.parent = parent;
	}

	cleanOne(convertPowers) {
		var hasDoneAnything = false;
		for (var i = 0; i < this.expressions.length; i++) {
			var child = this.expressions[i];
			if (child.getType() == CleaningExpressionType.INTEGER && child.asInt() == 0) {
				this.expressions.splice(i--, 1);
				hasDoneAnything = true;
			}
		}
		if (this.expressions.length == 1) {
			this.parent.replaceChild(this, this.expressions[0]);
			return true;
		} else if (this.expressions.length == 0) {
			this.parent.replaceChild(this, new CleaningInteger(0));
			return true;
		}
		for (var i = 0; i < this.expressions.length; i++) {
			var exp = this.expressions[i];
			hasDoneAnything = exp.cleanOne(convertPowers) || hasDoneAnything;
		}
		return hasDoneAnything;
	}

	cleanTwo() {
		var hasDoneAnything = false;
		for (var i = 0; i < this.expressions.length; i++) {
			var exp = this.expressions[i];
			hasDoneAnything = exp.cleanTwo() || hasDoneAnything;
		}
		return hasDoneAnything;
	}

	cleanThree() {
		var hasDoneAnything = false;
		for (var i = 0; i < this.expressions.length; i++) {
			var exp = this.expressions[i];
			hasDoneAnything = exp.cleanThree() || hasDoneAnything;
		}
		var logs = [];
		var negLogs = [];
		for (var i = 0; i < this.expressions.length; i++) {
			var child = this.expressions[i];
			if (child.getType() == CleaningExpressionType.LOG) {
				logs.push(child);
			} else if (child.getType() == CleaningExpressionType.NEGATE
					&& child.child.getType() == CleaningExpressionType.LOG) {
				negLogs.push(child);
			}
		}
		var toRemove = [];
		var newChild = null;
		for (var i = 0; i < logs.length; i++) {
			var log = logs[i];
			var matchingLogs = [];
			var matchingNegLogs = [];
			var base = log.getBase();
			for (var j = 0; j < logs.length; j++) {
				var otherLog = logs[j];
				if (otherLog.getBase().equals(base) && otherLog != log) {
					matchingLogs.push(otherLog);
				}
			}
			for (var j = 0; j < negLogs.length; j++) {
				var negLog = negLogs[j];
				if (negLog.child.getBase().equals(base)) {
					matchingNegLogs.push(negLog);
				}
			}
			if (matchingLogs.length + matchingNegLogs.length > 1) {
				var mult = new CleaningMultiply();
				toRemove.push(... matchingLogs);
				toRemove.push(... matchingNegLogs);
				for (var j = 0; j < matchingLogs.length; j++) {
					var l = matchingLogs[j];
					mult.addExpression(l.getChild());
				}
				for (var j = 0; j < matchingNegLogs.length; j++) {
					var l = matchingNegLogs[j];
					mult.addExpression(new CleaningInvert(l.child.getChild()));
				}
				newChild = new CleaningLog(base, mult);
				break;
			}
		}
		if (newChild == null) {
			for (var i = 0; i < negLogs.length; i++) {
				var negate = negLogs[i];
				var matchingNegLogs = [];
				var base = negate.child.getBase();
				for (var j = 0; j < negLogs.length; j++) {
					var negLog = negLogs[j];
					if (negLog.child.getBase().equals(base)) {
						matchingNegLogs.push(negLog);
					}
				}
				if (matchingNegLogs.length > 1) {
					var mult = new CleaningMultiply();
					toRemove.push(... matchingNegLogs);
					for (var j = 0; j < matchingNegLogs.length; j++) {
						var l = matchingNegLogs[j];
						mult.addExpression(l.child.getChild());
					}
					newChild = new CleaningNegate(new CleaningLog(base, mult));
					break;
				}
			}

		}
		if (newChild != null) {
			hasDoneAnything = true;
			newChild.setParent(this);
			this.expressions = this.expressions.filter(e=>!toRemove.includes(e));
			this.expressions.push(newChild);
		}
		return hasDoneAnything;
	}
	
	cleanFour() {
		var didSomething = false;
		var totalIntVal = 0;
		var reals = [];
		var negReals = [];
		for (var i = 0; i < this.expressions.length; i++) {
			var child = this.expressions[i];
			if (child.getType() == CleaningExpressionType.INTEGER) {
				totalIntVal += child.asInt();
				this.expressions.splice(i--, 1);
			} else if (child.getType() == CleaningExpressionType.REAL) {
				reals.push(child);
				this.expressions.splice(i--, 1);
			} else if (child.getType() == CleaningExpressionType.MULTIPLY && child.getExpressions().length == 2) {
				if (child.getExpressions()[0].getType() == CleaningExpressionType.INTEGER
						&& child.getExpressions()[1].getType() == CleaningExpressionType.REAL) {
					var num = child.getExpressions()[0].asInt();
					for (var j = 0; j < num; j++) {
						reals.push(child.getExpressions()[1]);
					}
					this.expressions.splice(i--, 1);
				} else if (child.getExpressions()[1].getType() == CleaningExpressionType.INTEGER
						&& child.getExpressions()[0].getType() == CleaningExpressionType.REAL) {
					var num = child.getExpressions()[1].asInt();
					for (var j = 0; j < num; j++) {
						reals.push(child.getExpressions()[0]);
					}
					this.expressions.splice(i--, 1);
				}
			} else if (child.getType() == CleaningExpressionType.NEGATE) {
				var cc = child.child;
				if (cc.getType() == CleaningExpressionType.INTEGER) {
					totalIntVal -= cc.asInt();
					this.expressions.splice(i--, 1);
				} else if (cc.getType() == CleaningExpressionType.REAL) {
					negReals.push(cc);
					this.expressions.splice(i--, 1);
				} else if (cc.getType() == CleaningExpressionType.MULTIPLY && cc.getExpressions().length == 2) {
					if (cc.getExpressions()[0].getType() == CleaningExpressionType.INTEGER
							&& cc.getExpressions()[1].getType() == CleaningExpressionType.REAL) {
						var num = cc.getExpressions()[0].asInt();
						for (var j = 0; j < num; j++) {
							negReals.push(cc.getExpressions()[1]);
						}
						this.expressions.splice(i--, 1);
					} else if (cc.getExpressions()[1].getType() == CleaningExpressionType.INTEGER
							&& cc.getExpressions()[0].getType() == CleaningExpressionType.REAL) {
						var num = cc.getExpressions()[1].asInt();
						for (var j = 0; j < num; j++) {
							negReals.push(cc.getExpressions()[0]);
						}
						this.expressions.splice(i--, 1);
					}
				}
			}
		}
		if (totalIntVal != 0) {
			if (totalIntVal > 0) {
				this.expressions.push(new CleaningInteger(totalIntVal));
			} else {
				this.expressions.push(new CleaningNegate(new CleaningInteger(-totalIntVal)));
			}
		}
		for (var i = 0; i < reals.length; i++) {
			var r = reals[i];
			if (negReals.includes(r)) {
				negReals.splice(negReals.indexOf(r), 1);
				reals.splice(i--, 1);
				didSomething = true;
			}
		}
		while (reals.length > 0) {
			var r = reals[0];
			reals.splice(0, 1);
			var num = 1;
			for (var i = 0; i < reals.length; i++) {
				if (reals[i].equals(r)) {
					reals.splice(i--, 1);
					num++;
					didSomething = true;
				}
			}
			for (var i = 0; i < negReals.length; i++) {
				if (negReals[i].equals(r)) {
					negReals.splice(i--, 1);
					num--;
					didSomething = true;
				}
			}
			if(Math.abs(num)==1){
				if (num > 0) {
					this.expressions.push(r);
				} else {
					var n = new CleaningNegate(r);
					n.setParent(this);
					this.expressions.push(n);
				}
			}else if (num != 0) {
				var mult = new CleaningMultiply();
				mult.addExpression(r);
				if (num > 0) {
					mult.addExpression(new CleaningInteger(num));
					mult.setParent(this);
					this.expressions.push(mult);
				} else {
					mult.addExpression(new CleaningInteger(-num));
					var n = new CleaningNegate(mult);
					n.setParent(this);
					this.expressions.push(n);
				}
			}
		}
		while (negReals.length > 0) {
			var r = negReals[0];
			negReals.splice(0, 1);
			var num = -1;
			for (var i = 0; i < negReals.length; i++) {
				if (negReals[i].equals(r)) {
					negReals.splice(i--, 1);
					num--;
					didSomething = true;
				}
			}
			if (Math.abs(num) == 1) {
				var n = new CleaningNegate(r);
				n.setParent(this);
				this.expressions.push(n);
			} else {
			var mult = new CleaningMultiply();
				mult.addExpression(r);
				mult.addExpression(new CleaningInteger(-num));
				var n = new CleaningNegate(mult);
				n.setParent(this);
				this.expressions.push(n);
			}
		}
		for (var i = 0; i < this.expressions.length; i++) {
			didSomething = this.expressions[i].cleanFour() || didSomething;
		}
		return didSomething;
	}

	toRendered(r) {
		var pos = "";
		var neg = "";
		var hasPrev = false;
		for (var exp of this.expressions) {
			if (exp.getType() == CleaningExpressionType.NEGATE) {
				neg+=r.minus();
				neg+=exp.child.toRendered(r);
			} else {
				if (hasPrev) {
					pos+=r.plus();
				}
				pos+=exp.toRendered(r);
				hasPrev = true;
			}
		}
		return pos + neg;
	}

	replaceChild(oldExp, newExp) {
		this.expressions.splice(this.expressions.indexOf(oldExp),1);
		this.expressions.push(newExp);
		newExp.setParent(this);
	}

	equals(obj) {
		if (!(obj instanceof CleaningAdd)) {
			return false;
		}
		if (obj.expressions.length != this.expressions.length) {
			return false;
		}
		if (obj.expressions.every(e=>this.expressions.includes(e))) {
			var exps = obj.expressions.slice(obj.expressions.length);
			for (var child of this.expressions) {
				if (!exps.includes(e=>e.equals(child))) {
					return false;
				}
				exps = exps.filter(e=>!e.equals(child));
			}
			return true;
		}
		return false;
	}

	canNegate() {
		for (var child of this.expressions) {
			if (child.getType() == CleaningExpressionType.NEGATE) {
				return true;
			}
		}
		return false;
	}

	negate() {
		var newExpressions = [];
		for (var child of this.expressions) {
			if (child.getType() == CleaningExpressionType.NEGATE) {
				newExpressions.push(child.child);
			} else {
				newExpressions.push(new CleaningNegate(child));
			}
		}
		for (var newExp of newExpressions) {
			newExp.setParent(this);
		}
		this.expressions = newExpressions;

	}

	asDouble() {
		var d = 0;
		for (var exp of this.expressions) {
			d += exp.asDouble();
		}
		return d;
	}

	clone() {
		var clone = new CleaningAdd();
		for (var child of this.expressions) {
			clone.addExpression(child.clone());
		}
		return clone;
	}
}
this.CleaningAdd = CleaningAdd;